src/main.rs

```diff
@@ -47,125 +96,265 @@ impl LogLevel {
-                if !output.status.success() {
-                    // combine and print stderr/stdout
-                    log::warn!("red script failed...");
-                }
-            }
-        }
-        log::warn!("SYSTEM is UNHEALTHY. Rolling back and rebooting...");
-        Command::new("rpm-ostree").arg("rollback").status()?;
-        reboot()?;
-        return Ok(());
-    }
-    for path in ["/etc/greenboot/green.d/*.sh"] {
-        for entry in glob(path)?.flatten() {
-            log::info!("running green check {}", entry.to_string_lossy());
-            let output = Command::new("bash").arg("-C").arg(entry).output()?;
-            if !output.status.success() {
-                // combine and print stderr/stdout
-                log::warn!("green script failed...");
-            }
-        }
+
+    if script_failure {
+        bail!("health-check failed!");
     }
     Ok(())
 }
 
-fn reboot() -> Result<(), Error> {
-    Command::new("systemctl").arg("reboot").spawn()?;
+/// runs the scripts in red.d when health-check fails
+fn run_red() -> Result<()> {
+    for path in GREENBOOT_INSTALL_PATHS {
+        let red_path = format!("{path}/red.d/");
+        if let Some(e) = run_scripts("red", &red_path).err() {
+            log::error!("red script runner error: {}", e.to_string());
+        }
+    }
     Ok(())
 }
 
-fn stamp() -> Result<(), Error> {
-    fs::create_dir_all("/etc/greenboot/")?;
-    File::create("/etc/greenboot/upgrade.stamp")?;
+/// runs the scripts green.d when health-check passes
+fn run_green() -> Result<()> {
+    for path in GREENBOOT_INSTALL_PATHS {
+        let green_path = format!("{path}/green.d/");
+        if let Some(e) = run_scripts("green", &green_path).err() {
+            log::error!("green script runner error: {}", e.to_string());
+        }
+    }
     Ok(())
 }
 
+/// triggers the diagnostics followed by the action on the outcome
+/// this also handles setting the grub variables and system restart
+fn health_check() -> Result<()> {
+    let config = GreenbootConfig::get_config();
+    log::debug!("{config:?}");
+    handle_motd("healthcheck is in progress")?;
+    match run_diagnostics() {
+        Ok(()) => {
+            log::info!("greenboot health-check passed.");
+            run_green()
+                .unwrap_or_else(|e| log::error!("cannot run green script: {}", e.to_string()));
+            handle_motd("healthcheck passed - status is GREEN")
+                .unwrap_or_else(|e| log::error!("cannot set motd: {}", e.to_string()));
+            handle_boot_status(true)?;
+            Ok(())
+        }
+        Err(e) => {
+            log::error!("Greenboot health-check failed!");
+            handle_motd("healthcheck failed - status is RED")
+                .unwrap_or_else(|e| log::error!("cannot set motd: {}", e.to_string()));
+            run_red().unwrap_or_else(|e| log::error!("cannot run red script: {}", e.to_string()));
```

Comment:
this else will never be triggered since you are not propagating the error

Reply:
propagated the error..